svm/src/transaction_processor.rs

```diff
@@ -228,20 +239,49 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
             sanitized_txs,
             check_results,
             error_counters,
-            &self.fee_structure,
             account_overrides,
             &program_cache_for_tx_batch.borrow(),
         );
         load_time.stop();
 
         let mut execution_time = Measure::start("execution_time");
 
+        // collect unique loaded accounts
+        let mut unique_loaded_accounts: HashMap<Pubkey, AccountSharedData> = HashMap::default();
+        for tx in loaded_transactions.iter().filter(|tx| tx.is_ok()) {
+            let accts = &tx.as_ref().unwrap().accounts;
+            for acct in accts.iter() {
+                unique_loaded_accounts.insert(acct.0, acct.1.clone());
+            }
+        }
+
+        let mut entry_level_dedup_map = HashSet::with_capacity(MAX_NUM_TRANSACTIONS_PER_BATCH);
+        let mut error = None;
+        let feature_set = callbacks.get_feature_set();
+        let rent_collector = callbacks.get_rent_collector();
         let execution_results: Vec<TransactionExecutionResult> = loaded_transactions
             .iter_mut()
             .zip(sanitized_txs.iter())
-            .map(|(load_result, tx)| match load_result {
+            .zip(check_results.iter())
+            .map(|((accts, tx), check_result)| match accts {
                 Err(e) => TransactionExecutionResult::NotExecuted(e.clone()),
                 Ok(loaded_transaction) => {
+                    let mut nonce = None;
+                    let mut lamports_per_signature = None;
+                    if let Ok(check_result) = check_result {
+                        nonce = check_result.nonce.clone();
+                        lamports_per_signature = check_result.lamports_per_signature;
+                    }
+
+                    let msg_hash = tx.message_hash();
+                    if entry_level_dedup_map.contains(msg_hash) {
+                        Self::update_loaded_accounts(loaded_transaction, &unique_loaded_accounts);
```

Comment:
The idea here is that if a transaction fails then its corresponding LoadedTransaction structure should be up-to-date according to the recent temp state. 

But, you are right, this could be done in the load_accounts so that it would error before execution. will push a commit.